quinn-proto/src/connection/mod.rs

```diff
@@ -541,7 +539,57 @@ where
         }
 
         if buf.is_empty() {
-            return None;
+            if !self.state.is_established() {
+                return None;
+            }
+
+            let space = &mut self.spaces[SpaceId::Data];
+            let probe_size = self.path.mtud.poll_transmit(space.next_packet_number)?;
+
+            // Send an MTU discovery probe
+            let exact_number = space.get_tx_number();
+            let number = PacketNumber::new(exact_number, space.largest_acked_packet.unwrap_or(0));
+            let header = Header::Short {
+                dst_cid: self.rem_cid,
+                number,
+                spin: if self.spin_enabled {
+                    self.spin
+                } else {
+                    self.rng.gen()
+                },
+                key_phase: self.key_phase,
+            };
+            let partial_encode = header.encode(&mut buf);
+            buf.write(frame::Type::PING);
+
+            let (header_crypto, packet_crypto) = if let Some(ref crypto) = space.crypto {
+                (&crypto.header.local, &crypto.packet.local)
+            } else {
+                unreachable!("must have keys in established state")
+            };
+
+            let tag_len = packet_crypto.tag_len();
```

Comment:
This repeats a bunch of code from the block above. Wrapping a cleaner abstraction around that feels non-trivial, but any thoughts on this would be welcome. Otherwise, I'd like to keep that as future follow-up.

Reply:
Hmm, I rather like this design where we only send MTU probes if we have nothing else to send and, since the MTU probe doesn't have any other frames, nothing is affected if the probe gets lost. Seems to me like packing ack-eliciting frames in the same packet invokes a substantial amount of (reasoning) complexity at relatively little benefit.

Reply:
> we only send MTU probes if we have nothing else to send and, since the MTU probe doesn't have any other frames, nothing is affected if the probe gets lost.

Bundling ACK and flow control frames into the probe doesn't undermine these concerns; it's data that we wouldn't have otherwise sent for the sake of efficiency, but we might as well throw it in if we're spending bandwidth on a probe anyway.

> packing ack-eliciting frames in the same packet

I don't quite follow. `PING` is ack-eliciting. If we didn't send ack-eliciting frames it wouldn't be a very good probe! In any case, the hope is that by going through the same path we can avoid the complexity and maintenance hazard of duplicated code.

Reply:
Bundling `ACK`s into a frame that is likely to get lost sounds potentially problematic, especially if we are using ACK decimation.